qiskit/pulse/channels/pulse_specification.py

```diff
@@ -0,0 +1,139 @@
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+# pylint: disable=invalid-name
+
+"""
+Specification of the device.
+"""
+from typing import List
+
+from .channels import AcquireChannel, MemorySlot, RegisterSlot
+from .pulse_channels import DriveChannel, ControlChannel, MeasureChannel
+from .qubit import Qubit
+
+
+class PulseSpecification:
+    """Implement a device specification."""
+
+    def __init__(self,
+                 n_drive: int,
+                 n_control: int,
+                 n_measure: int,
+                 n_registers: int,
+                 buffer: int = 0,
+                 hamiltonian=None):
+        """
+        Create pulse specification with number of channels.
+
+        Args:
+            n_drive: Number of drive channels.
+            n_control: Number of control channels.
+            n_measure: Number of acquire channels. Acquire channel is bound to a single memory slot.
+            n_registers: Number of classical registers.
+            buffer: Buffer that should be placed between instructions on channel.
+            hamiltonian (Hamiltonian): Hamiltonian object to extract device topology.
+        """
+        self._drives = [DriveChannel(idx, buffer) for idx in range(n_drive)]
+        self._controls = [ControlChannel(idx, buffer) for idx in range(n_control)]
+        self._measures = [MeasureChannel(idx, buffer) for idx in range(n_measure)]
+        self._acquires = [AcquireChannel(idx, buffer) for idx in range(n_measure)]
+        self._mem_slots = [MemorySlot(idx) for idx in range(n_measure)]
+        self._reg_slots = [RegisterSlot(idx) for idx in range(n_registers)]
+        self._qubits = []
+
+        if hamiltonian:
+            # TODO: create qubit from hamiltonian
+            pass
+
+    @classmethod
+    def from_device(cls, backend):
+        """
+        Create pulse specification with values from backend.
+
+        Args:
+            backend (BaseBackend): Backend configuration.
+
+        Returns:
+            PulseSpecification: New PulseSpecification configured by backend.
+        """
+        configuration = backend.configuration()
+        defaults = backend.defaults()
+
+        # TODO: define n_drive and n_acquire explicitly in backend
+        n_drive = len(configuration.qubit_lo_range)
+        n_control = configuration.n_uchannels
+        n_measure = len(configuration.meas_lo_range)
+        n_registers = configuration.n_registers
+        buffer = defaults.buffer
+        hamiltonian = configuration.hamiltonian
+
+        return PulseSpecification(n_drive=n_drive, n_control=n_control,
+                                  n_measure=n_measure, n_registers=n_registers,
+                                  buffer=buffer, hamiltonian=hamiltonian)
+
+    def __eq__(self, other):
```

Comment:
I'm not really sure of the use case for checking equality of two device specification? This might be adding too many features. It might be better to re-add at a later date if necessary.

Reply:
If we have some devices with the same topology, we don't need to recreate schedule. If we want to execute a schedule in parallel with multiple devices, we may want to check if the devices have the same topology. But I think such situation doesn't happen soon and we can remove equality check.